### PR TITLE
log important config of apiserver, observer, & scheduler

### LINKED ISSUE #1918: components should dump important configuration to the logs at startup
As I start my efforts to diagnose some outstanding bugs, I'm finding this would be extremely useful. It would be great to know when a component is perhaps not configured the way you thought it was.

Let's apply this to the apiserver, scheduler, and observer components for now.

Great care should be taken to only display _non-sensitive_ configuration -- no secretes.

### PR DESCRIPTION
**Pull request by AnuragThePathak**
Signed-off-by: Anurag Pathak <anuragpathak911@gmail.com>
Fixes #1918 

**What this PR does / why we need it**: I'm simply logging configuration structs with it's fields. As a result braces will also get printed but maybe it'll help to identify similar configurations by clubbing them.

**Special notes for your reviewer**: I have skipped quite a few configurations in api server as I was not sure if it will be useful to log them. Kindly let me know if you feel logging them would be useful.

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by krancour** [comment]
> I'm simply logging configuration structs with it's fields. As a result braces will also get printed but maybe it'll help to identify similar configurations by clubbing them.

For the API server, it looks like this:

```
{BrigadeID:brigade.brigade APIAddress:http://brigade-apiserver.brigade.svc.cluster.local/ GitInitializerImage:localhost:50108/brigadecore_brigade2-git-initializer:tilt-498fd491ab5ec9cc GitInitializerImagePullPolicy:IfNotPresent GitInitializerWindowsImage:brigadecore/brigade2-git-initializer-windows:v0.1.0 GitInitializerWindowsImagePullPolicy:IfNotPresent DefaultWorkerImage:localhost:50108/brigadecore_brigade2-worker:tilt-8a09a4ef2e05abec DefaultWorkerImagePullPolicy:IfNotPresent WorkspaceStorageClass:nfs NodeSelectorKey: NodeSelectorValue: TolerationKey: TolerationValue:}
```

That's not human-friendly, and the point of this is to help humans troubleshoot.

For the scheduler and observer, I see time intervals expressed in nanoseconds. Again -- not human friendly. I had to pull out my calculator to see that 86,400,000,000,000 nanoseconds is 24h.

For any of this to serve its intended purpose, all the output needs to be much more nicely and deliberately formatted.

There's also a major risk in indiscriminately dumping entire objects to the logs in this manner. You're not _currently_ displaying any sensitive information, but if a sensitive field is added to an existing type in the future, it will be too easy to forget that this information is being logged at startup and that will create problems.

> I have skipped quite a few configurations in api server as I was not sure if it will be useful to log them. Kindly let me know if you feel logging them would be useful.

I think there are a lot more config objects that would be useful to see. And be careful, because some of them _do_ already contain sensitive fields. So, not to belabor the point, but just dumping entire objects to the logs isn't going to work out well for those -- which is a strong indicator that the strategy isn't viable even for the things we're already logging.

Looking at the existing code more closely, there are also cases where we don't return a simple config object, but rather act on env vars to immediately create and return some component -- like a database connection. You're not going to dump a database connection object to the logs -- but I still might want to know the address of the database server (unless it has a password embedded in it -- which is possible).

Bottom line is that I think this issue isn't one that's solved with just a few lines of code. Some careful thought is required on how to approach this and things should be nicely formatted for ease of human consumption.

**Comment by AnuragThePathak** [comment]
Cool, I'll try to find some better way to get the job done. Maybe using a map can be a good idea.

**Comment by krancour** [comment]
@AnuragThePathak I've been thinking about this a lot more and a good option might be to introduce a leveled logger -- which is something we probably would do eventually anyway -- and use that to write debug level messages as the settings are read from their corresponding environment variables. And then this would also have the advantage that the extra info can be quieted when not actively troubleshooting.

**Comment by krancour** [comment]
I take back what I said about leveled logging. I've researched it a bit just now and there are a lot of new leveled loggers out there since the last time I used one. I don't want to rush that decision _and_ whatever we decide on should probably have a standardized configuration rolled into [brigade-foundations](https://github.com/brigadecore/brigade-foundations) so that it's easily reusable across all of Brigade's peripherals -- gateways, etc.

For now, let's stick with the idea of logging important configuration values _as they're read_ -- it just won't be leveled is all.

**Comment by krancour** [comment]
Going in this direction should isolate your changes to each component's `config.go` file.

**Comment by AnuragThePathak** [comment]
> For now, let's stick with the idea of logging important configuration values as they're read -- it just won't be leveled is all.

No issues. 
Regarding leveled loggers, I've used the uber-go/zap logger with jackc/pgx driver but haven't used that directly.

**Comment by krancour** [comment]
@AnuragThePathak zap currently looks like the best option. I just don't want to rush that decision because it will have project-wide consequences.

**Comment by AnuragThePathak** [comment]
I just realized that my first commit in this PR had a typo. Moreover, in accordance with current strategy we don't need that now, hence can be dropped. But wouldn't rebasing result in corrupting all other commits which were not not by me but prior to the one I'm talking.

**Comment by AnuragThePathak** [comment]
@krancour Updated the commits properly without affecting prior ones which were not created by me. BTW back to home now and expecting to be able to work properly on other issues.

File: v2/apiserver/config.go [hunk 1]
```diff
@@ -106,47 +109,62 @@ func substrateConfig() (kubernetes.SubstrateConfig, error) {
...
 	}
+	log.Println("DEFAULT_WORKER_IMAGE_PULL_POLICY: ",
+		config.DefaultWorkerImagePullPolicy)
 	config.DefaultWorkerImagePullPolicy =
 		api.ImagePullPolicy(defaultWorkerImagePullPolicyStr)
 	config.WorkspaceStorageClass, err =
 		os.GetRequiredEnvVar("WORKSPACE_STORAGE_CLASS")
+	if err == nil {
+		log.Println("WORKSPACE_STORAGE_CLASS: ", config.WorkspaceStorageClass)
+	}
```
[the remark this answers is not in the document]
Comment (krancour) [reply v2/apiserver/config.go:167]: And then you can `return config, nil` at the end of the function since nothing else between here and there could cause an error.

File: v2/apiserver/config.go [hunk 2]
```diff
@@ -244,16 +268,23 @@ func sessionsServiceConfig() (api.SessionsServiceConfig, error) {
...
+	log.Println("USER_SESSION_TTL: ", config.UserSessionTTL)
 	config.AdminUserIDs =
 		os.GetStringSliceFromEnvVar("ADMIN_USER_IDS", []string{})
+	log.Println("ADMIN_USER_IDS: ", config.AdminUserIDs)
 	config.GrantReadOnInitialLogin, err =
 		os.GetBoolFromEnvVar("GRANT_READ_ON_INITIAL_LOGIN", false)
+	if err == nil {
+		log.Println("GRANT_READ_ON_INITIAL_LOGIN: ", 
+		config.GrantReadOnInitialLogin)
+	}
```
Comment (krancour) [inline v2/apiserver/config.go:287]: I leave it up to you whether you want to apply the same advice from [this comment](https://github.com/brigadecore/brigade/pull/1945/files?diff=split&w=0#r873838015) here. It seems less of an issue here because we're so close to the bottom of the function and the return that it's very clear to see what's going on here without any changes. Totally up to you whether to change it or leave it like this.

**Comment by krancour** [comment]
Very minor corrections needed... otherwise LGTM.

**Comment by AnuragThePathak** [comment]
I chose to follow the pattern we've been following in the 2nd case as well.

**Review by krancour — APPROVED** [review]
